pkg/controllers/infracluster/infracluster_controller.go

```diff
@@ -0,0 +1,288 @@
+	log := ctrl.LoggerFrom(ctx).WithName("InfraClusterController")
+
+	res, err := r.reconcile(ctx, log)
+	if err != nil {
+		if err := r.setDegradedCondition(ctx, log); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set conditions for InfraCluster controller: %w", err)
+		}
+		return ctrl.Result{}, fmt.Errorf("error during reconcile: %w", err)
+	}
+
+	if err := r.setAvailableCondition(ctx, log); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set conditions for InfraCluster controller: %w", err)
+	}
+
+	return res, nil
+}
+
+// reconcile performs the main business logic for installing Cluster API components in the cluster.
+// Notably it fetches CAPI providers "transport" ConfigMap(s) matching the required labels,
+// it extracts from those ConfigMaps the embedded CAPI providers manifests for the components
+// and it applies them to the cluster.
+//
+//nolint:unparam
+func (r *InfraClusterController) reconcile(ctx context.Context, log logr.Logger) (ctrl.Result, error) {
+	switch r.Platform {
+	case configv1.AWSPlatformType:
+		if err := r.ensureAWSCluster(ctx); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error ensuring AWSCluster: %w", err)
+		}
+	case configv1.GCPPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.PowerVSPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.VSpherePlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.OpenStackPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	default:
+		log.Info(fmt.Sprintf("detected platform %q is not supported, skipping capi controllers setup", r.Platform))
+	}
+
+	// At this point, the InfraCluster exists.
+	managedByAnnotationVal, found := r.InfraCluster.GetAnnotations()[clusterv1.ManagedByAnnotation]
+	if !found {
+		// Could not find the managedByAnnotation on the InfraCluster object.
+		// This means, by definition, that the object is directly managed by CAPI.
+		// No action should be taken by this controller.
+		return ctrl.Result{}, nil
+	}
+
+	switch managedByAnnotationVal {
+	case managedByAnnotationValueClusterCAPIOperator:
+		// At this point it is this controller's responsibility to manage this InfraCluster object.
+		isReady, err := getReadiness(r.InfraCluster)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to get readiness for InfraCluster: %w", err)
+		}
+		if isReady {
+			// The Infrastructure for this CAPI Cluster is already ready - nothing to do.
+			return ctrl.Result{}, nil
+		}
+
+		infraClusterCopy, ok := r.InfraCluster.DeepCopyObject().(client.Object)
+		if !ok {
+			return ctrl.Result{}, fmt.Errorf("unable to convert InfraCluster runtime.Object to client.Object")
+		}
```

Comment:
If the infraCluster was returned by the `ensure` function, you could have it as a `client.Object` without the need for any additional deepcopying, since CR always returns copies anyway

Reply:
👍